zeekclient/brokertypes.py

```diff
@@ -0,0 +1,1145 @@
+                     for key in sorted(self._elements)]
+        }
+
+    @classmethod
+    def from_broker(cls, data):
+        res = Table()
+        for elem in data['data']:
+            res._elements[from_broker(elem['key'])] = from_broker(elem['value'])
+        return res
+
+
+# ---- Special types ---------------------------------------------------
+
+class ZeekEvent(Vector):
+    """Broker's event representation, as a vector of vectors.
+
+    This specialization isn't an official type in Broker's hierarchy: there's no
+    distinguishing @data-type for it. Zeek events are a specific interpretation
+    of nested vectors.
+
+    See Broker's websockets docs for an example:
+
+    https://docs.zeek.org/projects/broker/en/current/web-socket.html#encoding-of-zeek-events
+    """
+    def __init__(self, name, *args):
+        super().__init__()
+
+        self.name = name.to_py() if isinstance(name, String) else str(name)
+        self.args = list(args) or [] # list here is to avoid tuple/list type confusion
+
+        for arg in self.args:
+            if not isinstance(arg, Type):
+                raise TypeError('ZeekEvent constructor requires brokertype arguments')
+
+    def to_broker(self):
+        return {
+            '@data-type': 'vector',
+            'data': [
+                {
+                    "@data-type": "count",
+                    "data": 1
+                },
+                {
+                    "@data-type": "count",
+                    "data": 1
+                },
+                {
+                    "@data-type": "vector",
+                    "data": [
+                        String(self.name).to_broker(),
+                        {
+                            "@data-type": "vector",
+                            "data": [arg.to_broker() for arg in self.args],
+                        },
+                    ],
+                },
+            ],
+        }
+
+    @classmethod
+    def from_vector(cls, vec):
+        """Special case for an existing Vector instance: recast as Zeek event."""
+        if not isinstance(vec, Vector):
+            raise TypeError('cannot convert non-vector to Zeek event')
+
+        if (not len(vec) == 3 or
+            not isinstance(vec[2], Vector) or
+            not len(vec[2]) == 2 or
+            not isinstance(vec[2][0], String) or
+            not isinstance(vec[2][1], Vector)):
+            raise TypeError('invalid vector layout for Zeek event')
+
+        name = vec[2][0].to_py()
+        args = vec[2][1]
+        return ZeekEvent(name, *args._elements)
+
+    @classmethod
+    def from_broker(cls, data):
+        name = data['data'][2]['data'][0]['data']
+        res = ZeekEvent(name)
+        for argdata in data['data'][2]['data'][1]['data']:
+            res.args.append(from_broker(argdata))
+        return res
+
+
+# ---- Message types ---------------------------------------------------
```

Comment:
I'm wondering if it would be worth it to move the message types out to this module into their own script/namespace. At some places where they are used, I found it a bit unintuitive to see them referred to as 'types' along with things like strings, counts, etc.

Reply:
Agreed, yeah. I was thinking that it's okay if this module is a little bit of a grabbag right now, for the same reason as above: if/when it becomes its own module, it'll clearly undergo a bunch of polish, and while it's just in `zeek-client` I'm okay with it. But let me know if you feel strongly — happy to change this.